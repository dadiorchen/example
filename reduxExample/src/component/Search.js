/* The component to input search condition : keyword and status */
//@flow
import React from "react";
import {searchModel} from '../model/Search.js'
import {connect} from 'react-redux'
import {type TypeState} from '../model/TypeState.js'


type Props = {
	status	: number,
	autoSearch	: boolean,
	changeStatus	: (status : number) =>  void,
	changeKeyword	: (keyword : string) => void,
}
export class Search extends React.Component<Props,{
	//state
	keyword	: string,
}>{
	constructor( props : Props){
		super(props);
		this.state = {
			keyword	: '',
		}
	}
	/********************** properties ************/
	/********************** react method ***********/
	/********************** component method *******/
	handleSearch = () => {
		this.props.changeKeyword(this.state.keyword)
	}

	handleChange = (e : any) => {
		this.setState({keyword:e.target.value},() => {
			//auto search
			if(this.props.autoSearch){
				this.handleSearch()
			}
		})
	}

	render(){
		console.warn('+++++render:search')
		return(
			<div
				style={{
					padding	: 5,
					background	: 'gray',
				}}
			>
				<div>
					<input type='text' onChange={this.handleChange} value={this.state.keyword} />
					<button
						onClick={this.handleSearch}
					>
					Search
					</button>
				</div>
				<div
					style={{
						margin	: 3,
						fontSize	: 19,
					}}
					className='search-status'
					onClick={() => {
						//change the status : -1 ,0 , 1  = ( a + 1 % 3 ) (-1,0)(0,1)(1,2)
						let newStatus = (this.props.status + 1 ) % 3
						if(newStatus === 2) newStatus = -1
						this.props.changeStatus(newStatus)
					}}
				>
					The status:
					<span
						style={{
							cursor	: 'pointer',
						}}
					>
					<strong>
					{this.props.status}
					</strong>
					</span>
				</div>
			</div>
		)
	}
}

const CSearch = require('../config.js').CONFIG.SANDBOX_RENDER() || 
	connect(
		(state : TypeState ) => {
			return {
				status	: state.search.status,
				autoSearch	: state.setting.autoSearch,
			}
		},
		{
			changeKeyword	: searchModel.actions.changeKeyword,
			changeStatus	: searchModel.actions.changeStatus,
		}
	)(Search)

export {CSearch}
