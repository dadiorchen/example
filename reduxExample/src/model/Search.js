//@flow
/* The model for search */
import todoModel from './TodoModel.js'

class SearchModel {
	reducers = {
		search	: (state : {keyword : string,status : number} = {keyword : '',status : -1},action:any) => {
			switch(action.type){
				case 'SEARCH_SET_STATUS' : {
					const newState = {...state}
					newState.status = action.status
					return newState
				}
				case 'SEARCH_SET_KEYWORD' : {
					const newState = {...state}
					newState.keyword = action.keyword
					return newState
				}
				default:
					return state
			}
		}
	}

	actions = {
		searchSetStatus	: (status : number) => {
			return {
				type	: 'SEARCH_SET_STATUS',
				status,
			}
		},
		searchSetKeyword	: (keyword : string) => {
			return {
				type	: 'SEARCH_SET_KEYWORD',
				keyword,
			}
		},
		changeStatus	: (status : number) => (dispatch : any , getState : any) => {
			dispatch(this.actions.searchSetStatus(status))
			//inform the todo model to change
			dispatch(todoModel.actions.todoFilterByStatus(status))
		},
		changeKeyword	: (keyword : string ) => (dispatch : any , getState : any) => {
			dispatch(this.actions.searchSetKeyword(keyword))
			dispatch(todoModel.actions.todoFilterByKeyword(keyword))
		},
	}
}

export const searchModel = new SearchModel()
