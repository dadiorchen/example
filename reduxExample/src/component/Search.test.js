//@flow
import React from 'react'
import {shallow} from 'enzyme'
import {Search} from './Search.js'
import toJson from 'enzyme-to-json'

describe('TestSearch',() =>{
	it('TestSearch',() => {
		const mock = jest.fn()
		const mockA = jest.fn()
		const w = shallow(<Search 
				autoSearch={false}
				changeKeyword={mock}
				status={-1}
				changeStatus={mockA}
			/>)
		expect(toJson(w)).toMatchSnapshot()
		//click
		w.find('button').simulate('click')
		expect(mock).toHaveBeenCalled()

		w.find('.search-status').simulate('click')
		expect(mockA).toHaveBeenCalled()
	})

	it('TestSearchAuto',() => {
		const mock = jest.fn()
		const mockA = jest.fn()
		const w = shallow(<Search 
				autoSearch={true}
				changeKeyword={mock}
				status={-1}
				changeStatus={mockA}
			/>)
		//onChange
		w.instance().handleChange({target:{value:'t'}})
		expect(mock).toHaveBeenCalled()

	})


})
