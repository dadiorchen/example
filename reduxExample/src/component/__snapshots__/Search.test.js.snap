// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestSearch TestSearch 1`] = `
<div
  style={
    Object {
      "background": "gray",
      "padding": 5,
    }
  }
>
  <div>
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      onClick={[Function]}
    >
      Search
    </button>
  </div>
  <div
    className="search-status"
    onClick={[Function]}
    style={
      Object {
        "fontSize": 19,
        "margin": 3,
      }
    }
  >
    The status:
    <span
      style={
        Object {
          "cursor": "pointer",
        }
      }
    >
      <strong>
        -1
      </strong>
    </span>
  </div>
</div>
`;
